File: model-client/src/main/java/org/modelix/model/api/PNodeReference.java
package org.modelix.model.api;

import org.jetbrains.annotations.Nullable;

public class PNodeReference implements INodeReference {
  private long id;

  public PNodeReference(long id) {
    this.id = id;
  }

  public long getId() {
    return id;
  }

  @Nullable
  @Override
  public INode resolveNode(INodeResolveContext context) {
    if (context instanceof PNodeResolveContext) {
      return new PNodeAdapter(id, ((PNodeResolveContext) context).getBranch());
    } else {
      return context.resolve(this);
    }
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || this.getClass() != o.getClass()) {
      return false;
    }

    PNodeReference that = (PNodeReference) o;
    if (id != that.id) {
      return false;
    }

    return true;
  }

  @Override
  public int hashCode() {
    int result = 0;
    result = 31 * result + (int) (id ^ (id >> 32));
    return result;
  }

  @Override
  public String toString() {
    return "PNodeReference{" + "id=" + id + "}";
  }
}

File: model-client/src/main/java/org/modelix/model/api/PNodeReference.kt
package org.modelix.model.api

class PNodeReference(val id: Long) : INodeReference {

    override fun resolveNode(context: INodeResolveContext?): INode? {
        return if (context is PNodeResolveContext) {
            PNodeAdapter(id, context.branch)
        } else {
            context!!.resolve(this)
        }
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || this.javaClass != o.javaClass) {
            return false
        }
        val that = o as PNodeReference
        return if (id != that.id) {
            false
        } else true
    }

    override fun hashCode(): Int {
        var result = 0
        result = 31 * result + (id xor (id shr 32)).toInt()
        return result
    }

    override fun toString(): String {
        return "PNodeReference{id=$id}"
    }

}